File: container_genetic_algorithm_solver/general_solver_functions_access.py
from logger import setup_logger

from aiohttp import ClientSession
from os import environ
import os

logger = setup_logger(logger_name=os.path.basename(__file__).split(".")[0])
api_key = str(environ["SOLVER_FUNCTIONS_KEY"])


async def calculate_board_fitness_report(
    board: list, zone_height: int, zone_length: int
) -> tuple:

    """Calculate Board Fitness Report

    This function uses the general solver functions api to calculate and return all the different collisions on a given board array
    representation.

    Args:
        board (list): A full filled board representation.
        zone_height (int): The zones height.
        zone_length (int): The zones length.

    Returns:
        int: Total collisions on the board.
        int: Total collisions on the board columns.
        int: Total collisions on the board rows.
        int: Total collisions on the board zones.
    """

    body = {"zoneHeight": zone_height, "zoneLength": zone_length, "board": board}
    url = str(environ["FITNESS_REPORT_SCORE_LINK"])
    response_body = dict()

    headers = {"Authorization": api_key, "Content-Type": "application/json"}
    async with ClientSession(headers=headers) as session:
        async with session.post(url=url, json=body) as response:
            response_body = await response.json()

    return (
        response_body["totalCollisions"],
        response_body["columnCollisions"],
        response_body["rowCollisions"],
        response_body["zoneCollisions"],
    )


async def calculate_board_fitness_single(
    board: list, zone_height: int, zone_length: int
) -> int:

    """Calculate Board Fitness Single

    This function uses the general solver functions api to calculate and return the total of all the collisions on a given board
    array representation.

    Args:
        board (list): A full filled board representation.
        zone_height (int): The zones height.
        zone_length (int): The zones length.

    Returns:
        int: Total collisions on the board.
    """

    body = {"zoneHeight": zone_height, "zoneLength": zone_length, "board": board}
    url = str(environ["FITNESS_SINGLE_SCORE_LINK"])
    response_body = dict()

    headers = {"Authorization": api_key, "Content-Type": "application/json"}
    async with ClientSession(headers=headers) as session:
        async with session.post(url=url, json=body) as response:
            response_body = await response.json()

    return response_body["fitnessScore"]


async def board_random_initialization(
    fixed_numbers_board: list, zone_height: int, zone_length: int
) -> list:

    """Board Random Initialization

    This function uses the general solver functions api to fill randomly a board based on its initial state, where just the fixed
    numbers are on the board, the white spaces need to be represented with a 0 and just the spaces with zero are changed for random
    numbers that are not in the board untill the board is filled.

    Args:
        fixed_numbers_board (list): A board representation that includes just the fixed numbers.
        zone_height (int): The zones height.
        zone_length (int): The zones length.

    Returns:
        list: A full filled board representation.
    """

    body = {
        "fixedNumbersBoard": fixed_numbers_board,
        "zoneHeight": zone_height,
        "zoneLength": zone_length,
    }
    url = str(environ["RANDOM_INITIALIZATION_LINK"])
    response_body = dict()

    headers = {"Authorization": api_key, "Content-Type": "application/json"}
    async with ClientSession(headers=headers) as session:
        async with session.post(url=url, json=body) as response:
            response_body = await response.json()

    return response_body["board"]


async def board_random_mutation(board: list, fixed_numbers_board: list) -> list:

    """Board Random Mutation

    This function uses the general solver functions api to mutate randomly a board based on its initial state, the mutation affect
    just the not fixed numbers on the board.

    Args:
        fixed_numbers_board (list): A board representation that includes just the fixed numbers.
        board (list): A full filled board representation.

    Returns:
        list: A full filled board representation with a mutation in one of its rows.
    """

    body = {"fixedNumbersBoard": fixed_numbers_board, "board": board}
    url = str(environ["RANDOM_MUTATION_LINK"])
    response_body = dict()

    headers = {"Authorization": api_key, "Content-Type": "application/json"}
    async with ClientSession(headers=headers) as session:
        async with session.post(url=url, json=body) as response:
            response_body = await response.json()

    return response_body["board"]

File: container_hill_climbing_solver_go/general_solver_functions_access.go
package main

import (
	"encoding/json"
	"net/http"
	"os"

	"github.com/ansel1/merry"
	"github.com/go-resty/resty/v2"
)

var apiKey string = os.Getenv("SOLVER_FUNCTIONS_KEY")

/*
   This function uses the general solver functions api to calculate and return all the different collisions on a given board array
   representation.
*/
func CalculateBoardFitnessReport(board [][]uint8, zoneHeight, zoneLength uint16) (totalCollisions, columnCollisions, rowCollisions, zoneCollisions uint16, err error) {

	var boardFitnessReportApiUrl string = os.Getenv("FITNESS_REPORT_SCORE_LINK")
	boardFitnessReportRequest := &BoardFitnessReportRequest{
		ZoneHeight: zoneHeight,
		ZoneLength: zoneLength,
		BoardArray: board,
	}

	boardFitnessReportResponse, err := resty.New().R().
		SetHeader("Content-Type", "application/json").
		SetHeader("Authorization", apiKey).
		SetBody(boardFitnessReportRequest).
		Get(boardFitnessReportApiUrl)

	if err != nil {
		return 0, 0, 0, 0, merry.Wrap(err).Append("unable to get the board fitness report").
			WithValue("response code", boardFitnessReportResponse.Status()).
			WithValue("request body", boardFitnessReportRequest).
			WithHTTPCode(http.StatusInternalServerError)
	}

	if !boardFitnessReportResponse.IsSuccess() {
		return 0, 0, 0, 0, merry.New("error while getting the board fitness report").
			WithValue("response code", boardFitnessReportResponse.Status()).
			WithValue("request body", boardFitnessReportRequest).
			WithHTTPCode(http.StatusInternalServerError)
	}

	var boardFitnessReportResponseRawBody []byte = boardFitnessReportResponse.Body()
	var boardFitnessReportResponseBody BoardFitnessReportResponse

	err = json.Unmarshal(boardFitnessReportResponseRawBody, &boardFitnessReportResponseBody)
	if err != nil {
		return 0, 0, 0, 0, merry.New("error while unmarshalling the board fitness response body").
			WithValue("response body", string(boardFitnessReportResponseRawBody)).
			WithHTTPCode(http.StatusInternalServerError)
	}

	return boardFitnessReportResponseBody.TotalCollisions,
		boardFitnessReportResponseBody.ColumnCollisions,
		boardFitnessReportResponseBody.RowCollisions,
		boardFitnessReportResponseBody.ZoneCollisions,
		nil

}

/*
   This function uses the general solver functions api to calculate and return the total of all the collisions on a given board
   array representation.
*/
func CalculateBoardFitnessSingle(board [][]uint8, zoneHeight, zoneLength uint16) (fitnessScore uint16, err error) {

	var boardFitnessSingleApiUrl string = os.Getenv("FITNESS_SINGLE_SCORE_LINK")
	boardFitnessReportRequest := &BoardFitnessSingleRequest{
		ZoneHeight: zoneHeight,
		ZoneLength: zoneLength,
		BoardArray: board,
	}

	boardFitnessSingleResponse, err := resty.New().R().
		SetHeader("Content-Type", "application/json").
		SetHeader("Authorization", apiKey).
		SetBody(boardFitnessReportRequest).
		Get(boardFitnessSingleApiUrl)

	if err != nil {
		return 0, merry.Wrap(err).Append("unable to get the board fitness score").
			WithValue("response code", boardFitnessSingleResponse.Status()).
			WithValue("request body", boardFitnessReportRequest).
			WithHTTPCode(http.StatusInternalServerError)
	}

	if !boardFitnessSingleResponse.IsSuccess() {
		return 0, merry.New("error while getting the board fitness score").
			WithValue("response code", boardFitnessSingleResponse.Status()).
			WithValue("request body", boardFitnessReportRequest).
			WithHTTPCode(http.StatusInternalServerError)
	}

	var boardFitnessSingleResponseRawBody []byte = boardFitnessSingleResponse.Body()
	var boardFitnessSingleResponseBody BoardFitnessSingleResponse

	err = json.Unmarshal(boardFitnessSingleResponseRawBody, &boardFitnessSingleResponseBody)
	if err != nil {
		return 0, merry.New("error while unmarshalling the board fitness response body").
			WithValue("response body", string(boardFitnessSingleResponseRawBody)).
			WithHTTPCode(http.StatusInternalServerError)
	}

	return boardFitnessSingleResponseBody.FitnessScore, nil

}
